### PR TITLE
*: Add DAG mode

### PR DESCRIPTION
**Pull request by zimulala**
PTAL @coocood @hanfei1991 @shenli

### CODE REVIEW

File: proto/executor.proto [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+message Selection {
+	// Where condition.
+	repeated Expr conditions = 1;   
+}
+
+message Aggregation {
+	// Group by clause.
+	repeated Expr group_by = 1;
+	// Gggregate functions.
```
Comment (hanfei1991) [inline proto/executor.proto]: Agg

File: proto/executor.proto [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+	repeated Expr agg_func = 2;
+}
+
+message TopN {
+	// Order by clause.
```
Comment (hanfei1991) [inline proto/executor.proto]: Order by + limit

File: proto/executor.proto [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+	optional int64 table_id = 1 [(gogoproto.nullable) = false];
+	optional int64 index_id = 2 [(gogoproto.nullable) = false];
+	repeated ColumnInfo columns = 3;
+	optional bool desc = 4;
+	// Disjoint handle ranges to be scanned.
+	repeated KeyRange ranges = 5; 
+}
+
+message Selection {
+	// Where condition.
```
Comment (hanfei1991) [inline proto/executor.proto]: Conditions

File: proto/expression.proto [hunk 4]
```diff
@@ -158,3 +158,9 @@ message Expr {
 	optional bytes val = 2;
 	repeated Expr children = 3;
 }
+
+// ByItem type for group by and order by.
```
Comment (hanfei1991) [inline proto/expression.proto]: Not for group by any more.

File: proto/select.proto [hunk 5]
```diff
@@ -117,3 +106,24 @@ message RowMeta {
     	optional int64 handle = 1 [(gogoproto.nullable) = false];
     	optional int64 length = 2 [(gogoproto.nullable) = false];
 }
+
+// DAGRequest works like a simplified select statement with DAG mode.
```
Comment (hanfei1991) [inline proto/select.proto]: DAG works no longer like a select Request

File: proto/executor.proto [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+option (gogoproto.sizer_all) = true;
+option (gogoproto.unmarshaler_all) = true;
+
+enum ExecType {
+	TypeTableScan = 0;
+	TypeIndexScan = 1;
+	TypeSelection = 2;
+	TypeAggregation = 3;
+	TypeTopN = 4;
+	TypeLimi = 5;
```
Comment (coocood) [inline proto/executor.proto]: TypeLimit

File: proto/executor.proto [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+	// Disjoint handle ranges to be scanned.
+	repeated KeyRange ranges = 4; 
+}
+
+message IndexScan {
+	optional int64 table_id = 1 [(gogoproto.nullable) = false];
+	optional int64 index_id = 2 [(gogoproto.nullable) = false];
+	repeated ColumnInfo columns = 3;
+	optional bool desc = 4;
+	// Disjoint handle ranges to be scanned.
```
Comment (coocood) [inline proto/executor.proto]: Index range is not handle ranges.

File: proto/executor.proto (hunk 2 above)
Comment (zimulala) [inline proto/executor.proto]: This comment is for `order_by`.

File: proto/expression.proto (hunk 4 above)
Comment (zimulala) [inline proto/expression.proto]: The structure of `SelectRequest` is still using this.

**Comment by zimulala** [comment]
PTAL @coocood @hanfei1991

**Comment by coocood** [comment]
LGTM

File: proto/executor.proto [hunk 8]
```diff
@@ -37,28 +37,24 @@ message TableScan {
...
-	repeated KeyRange ranges = 4; 
 }
 
 message IndexScan {
 	optional int64 table_id = 1 [(gogoproto.nullable) = false];
 	optional int64 index_id = 2 [(gogoproto.nullable) = false];
 	repeated ColumnInfo columns = 3;
 	optional bool desc = 4;
-	// Disjoint handle ranges to be scanned.
-	repeated KeyRange ranges = 5;
```
Comment (hanfei1991) [inline proto/executor.proto]: It's better not to remove this.

**Comment by hanfei1991** [comment]
LGTM

File: proto/executor.proto (hunk 8 above)
Comment (zimulala) [inline proto/executor.proto]: This field has existed in the Structure of `Request`.

**Pull request merged**